### PR TITLE
Ready for review

### PR DESCRIPTION
**Pull request by noomdalv**

### CODE REVIEW

File: enum.rb
Comment (Ispirett) [inline enum.rb]: You are missing and end statement at the end of the file

File: enum.rb
Comment (Ispirett) [inline enum.rb]: my_all? should be able to a parameter so for example array.my_all?(4)

File: enum.rb
Comment (Ispirett) [inline enum.rb]: my_any? should be able to a parameter so for example array.my_any?(4)

File: enum.rb
Comment (Ispirett) [inline enum.rb]: my_none? should be able to a parameter so for example array.my_none?(String)

File: enum.rb
Comment (Ispirett) [inline enum.rb]: My count is not working as expected you are getting a local jump error .use **return true if !block_given?** This should be used as a guard case in some of your methods.

**Review by Ispirett — CHANGES_REQUESTED** [review]
Hey, @noomdalv great attempt so far  :100: :1st_place_medal:  but your methods are not giving the same output as original methods.

please do bug :bug:  fixing and submit for another review.

## Git reminder.

Friendly reminder: You should use Git Flow in this project. However, changing it now would take too much time.
Please make sure you use Git Flow in the next project. You will find instructions in the Patwright lesson description.

**Comment by noomdalv** [comment]
Corrections done:

- [x]  End statement implemented correctly in #my_count method

- [x] These methods should receive a block and return either true or false based on the block condition, test cases added in comments at end of the file for these methods:

- #my_all?
- #my_any?
- #my_none? 

- [x] #my_count method should return the correct number of elements, test case added.

**Comment by Torres-ssf** [comment]
### Pull Request Under Review By Sérgio (TSE)

File: enum.rb
Comment (Torres-ssf) [inline enum.rb:4]: [REQUIRED]

- [ ] my_each must return an Enumerator if no block is given.

File: enum.rb
Comment (Torres-ssf) [inline enum.rb:16]: [REQUIRED]

- [ ] my_each_with_index must return an Enumerator if no block is given.

File: enum.rb
Comment (Torres-ssf) [inline enum.rb]: [REQUIRED]

- [ ] my_all? it`s giving an error when a block is passed. Your method should behave like the original one

```
%w[ant bear cat].all? { |word| word.length >= 3 } #=> true
%w[ant bear cat].all? { |word| word.length >= 4 } #=> false
%w[ant bear cat].all?(/t/)                        #=> false
[1, 2i, 3.14].all?(Numeric)                       #=> true
[nil, true, 99].all?                              #=> false
[].all?                                           #=> true
```

File: enum.rb
Comment (Torres-ssf) [inline enum.rb]: [REQUIRED]

- [ ] my_none? it`s also giving an error when a block is passed. Your method should behave like the original one: 

```
%w{ant bear cat}.none? { |word| word.length == 5 } #=> true
%w{ant bear cat}.none? { |word| word.length >= 4 } #=> false
%w{ant bear cat}.none?(/d/)                        #=> true
[1, 3.14, 42].none?(Float)                         #=> false
[].none?                                           #=> true
[nil].none?                                        #=> true
[nil, false].none?                                 #=> true
[nil, false, true].none?                           #=> false
```

File: enum.rb
Comment (Torres-ssf) [inline enum.rb]: [REQUIRED]
my_count must:

- [ ] Count the number of items in enum that are equal to item if an argument is given.

- [ ] Count the number of elements yielding a true value if a block is given

Should behave like the original:
```
ary = [1, 2, 4, 2]
ary.count               #=> 4
ary.count(2)            #=> 2
ary.count{ |x| x%2==0 } #=> 3
```

File: enum.rb
Comment (Torres-ssf) [inline enum.rb:97]: [REQUIRED]

- [ ] Should return an Enumerator if no block is given.

- [ ] Use only one my_map

File: enum.rb
Comment (Torres-ssf) [inline enum.rb]: [REQUIRED]

- [ ] You don't need this method. You can delete it.

File: enum.rb
Comment (Torres-ssf) [inline enum.rb]: [REQUIRED]

- [ ] You don't need this method. You can delete it.

File: enum.rb
Comment (Torres-ssf) [inline enum.rb]: [REQUIRED]

- [ ] Should behave like the original. Check this [link](https://apidock.com/ruby/Enumerable/inject)

- [ ] When a symbol is specified my_inject combines each element of the collection by applying the symbol as a named method

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
### Hi there mate :smile: 
Nice job implementing these changes :+1: :clap: 
### But we still have changes to make :muscle: 
These changes are required to get your project approved.

### One important thing. You select the wrong project when were filling the request review form. You selected "Testing Your Ruby Code - Enumerable". The right one is "Advanced Building Blocks - Enumareble Methods". Please make sure to select this one next time.

[FOR FUTURE PROJECT]
 You should use Git Flow in future projects. It's really easy to use it and will make your repository looks more professional. You will find instructions in the Patwright lesson description. 

Please kindly make these changes and submit for code review again. Happy coding :computer:

**Comment by noomdalv** [comment]
Corrections:
- #my_each and #my_each_with_index should return an enumerator if not block given
- #my_all and #my_none should now accept blocks correctly
- #my count should now accept parameters
- #my_injection can now accept symbols to specify the method applied to the accumulator

File: enum.rb
Comment (maelfosso) [inline enum.rb:83]: Sorry there is  a bug here
```ruby
irb(main):024:0> %w{ant bear cat}.none? { |word| word.length >= 4 }
=> false
irb(main):025:0> %w{ant bear cat}.my_none? { |word| word.length >= 4 }
=> true
```

Comment (maelfosso) [reply enum.rb:83]:
```ruby
irb(main):032:0> [nil, false, true].none?
=> false
irb(main):033:0> [nil, false, true].my_none?
=> true
```

**Review by maelfosso — CHANGES_REQUESTED** [review]
Hello,

Thank you for the improvements :clap: 
You're Just at one step from the end.
Please take a look at the comments below.

Kindly, make those changes and submit a new code review request.
Thank you.

File: enum.rb
Comment (samgaco) [inline enum.rb:9]: it was required by @Torres-ssf that : 

>  my_each must return an Enumerator if no block is given.


you can check that if we do 
```
irb(main):001:0> [1,2,3].each
=> #<Enumerator: [1, 2, 3]:each>
```
we get an enumerator, testing your method we get an array of numbers

File: enum.rb
Comment (samgaco) [reply enum.rb:16]: the same requirement as for **my_each** still not met

File: enum.rb
Comment (samgaco) [inline enum.rb:21]: >  my_each_with_index must return an Enumerator if no block is given.

File: enum.rb
Comment (samgaco) [reply enum.rb:83]: :heavy_check_mark:

File: enum.rb
Comment (samgaco) [inline enum.rb:101]: >  Should return an Enumerator if no block is given.


what we get now if we excute `[1,2,3].my_map` is:


`enum.rb:96:in `my_map': no block given (yield) (LocalJumpError)`

**Review by samgaco — CHANGES_REQUESTED** [review]
Hi @noomdalv :high_brightness: 

I see you have met all requirements, but still left some to implement requested in the second review. I have commented them below. Please :pray: work on those last changes and submit back for another code review once you are done.

------------------------
moving on :arrow_down: 
* **[For your next code review requests]**  Add a short summary of changes in Pull Request description
------------------------
* ###### STATUS OF THE PROJECT: REQUESTED CHANGES :stop_sign:

**Comment by noomdalv** [comment]
Corrections done
These methods should work properly now and return enumerator if no block is given:
#my_each 
#my_each_with_index 
#my_map

File: enum.rb
Comment (bolah2009) [inline enum.rb:70]: #### :bug: 

```ruby
irb(main):005:0> require_relative 'enum'
#<Enumerator: [1, 2, 3]:map>
=> true
irb(main):006:0> [1,2,3,4].my_none?
=> true
irb(main):007:0> [1,2,3,4].none?
=> false
irb(main):008:0> [1,2,3,4].none?(4)
=> false
irb(main):009:0> [1,2,3,4].my_none?(4)
=> true
irb(main):010:0> [1,2,3,4].none?{|i| i> 5}
=> true
irb(main):011:0> [1,2,3,4].my_none?{|i| i> 5}
=> false
```

### `my_none?` is not working as expected.
Please [see here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F) for expected behavior.

File: enum.rb
Comment (bolah2009) [inline enum.rb:51]: Let's add one more case: when a value other than `Regexp` or `Class` is passed.

```ruby
irb(main):002:0> [1,2,3,4,5].all?(3)
=> false
irb(main):003:0> [1,2,3,4,5].my_all?(3)
=> true
```

#### :bulb: You can use `else` as follows: :point_down: 

```suggestion
    else
      my_each { |x| return false unless x == arg }
    end
```

File: enum.rb
Comment (bolah2009) [inline enum.rb:31]: #### :bug: 

Like `my_each` please let's handle cases when no block is given.

```ruby
irb(main):002:0> [1,2,3,4,5].select
=> #<Enumerator: [1, 2, 3, 4, 5]:select>
irb(main):003:0> [1,2,3,4,5].my_select
LocalJumpError (no block given (yield))
```

```suggestion
    return to_enum unless block_given?

    i = 0
```

File: enum.rb
Comment (bolah2009) [inline enum.rb]: Please, let's abstain from using `p`, `puts` and others in production code. :blush:

File: enum.rb
Comment (bolah2009) [inline enum.rb]: Let's delete this `p` too.

File: enum.rb
Comment (bolah2009) [inline enum.rb]: Let's implement `my_any?` like `my_all?`
It should be able to accept arguments other than a block.

```ruby
irb(main):004:0> [1,2,3,4].any?
=> true
irb(main):005:0> [1,2,3,4].my_any?
LocalJumpError (no block given (yield))
irb(main):006:0> [1,2,3,4].any?(2)
=> true
irb(main):007:0> [1,2,3,4].my_any?(2)
ArgumentError (wrong number of arguments (given 1, expected 0))
irb(main):008:0> [1,2,3,4].any?(Integer)
=> true
irb(main):009:0> [1,2,3,4].my_any?(Integer)
ArgumentError (wrong number of arguments (given 1, expected 0))
```

#### See [`#any` in ruby doc here](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F)

File: enum.rb
Comment (bolah2009) [inline enum.rb]: There's a typo here: `to_enum` not `enum`

**Review by bolah2009 — CHANGES_REQUESTED** [review]
## Awesome Job @noomdalv 

### Nice improvement, you almost there. :smile: 

Check out the comments below, I also include some hints. 

Feel free to reach me on slack if you have questions

Please fix this and submit another code review request.

**Comment by noomdalv** [comment]
5th review done, list of changes:
#my_select corrected, should return enumerator if no block is given
#my_none corrected, return statements were reversed, it should work properly
#my_all corrected, added case to evaluate objects other than array/hash
#my_map corrected(typo)
#my_each corrected, removed unnecesary return
#my_each_with_index corrected, removed unnecesary return

File: enum.rb
Comment (mcrd25) [inline enum.rb:109]: I feel like maybe you've overcomplicated your my_inject implementation.

Your `my_inject` does not quite work as should with block and symbol.

How `my_inject` works: there are two possible arguments that can be passed. Either the initial accumulator or a symbol or both which you understand.
Here are the following conditions to consider:
- First argument is nil or symbol then the initial accumulator is set to the first element in the array.
- if a block is given and arg1 has value (not symbol) then you yield the block for each element in an array
- if block is given and arg1 is nil then yield block `|num| accumulator = yield(accumulator, num)` from the second element in array `[1..-1].my_each `
- If first arg1 is symbol then you do `|num| accumulator = accumulator.send(arg1, num)` for each element in array from the second element in the array (`[1..-1]`)
- if arg2 is a symbol then you do `|num| accumulator = accumulator.send(arg1, num)` for each element in array (from first)

You return the accumulator at the end of the method. The accumulator is the only thing you'd need to return for this method.

Follow this logic and you shall be fine and pass all tests. 👍

File: enum.rb
Comment (mcrd25) [inline enum.rb]: As it'd give false positive false results as the first element may be false but that does not mean the next is not true. It would however still return false.

File: enum.rb
Comment (mcrd25) [inline enum.rb:70]: `my_none` is simply opposite of `my_any` So once you've got `my_any` working you'd simply have to negate `my_any`

File: enum.rb
Comment (mcrd25) [inline enum.rb:58]: [OPTIONAL]
a simple `===` would cover **Class**, **RegExp** and **Pattern**
All of the extra ifs not necessary. Only need 3
For block, when arg not nil and when arg nil and block not given.

File: enum.rb
Comment (mcrd25) [inline enum.rb:55]: Your `my_any` does not quite work as should.
Two considerations:
- When no block or argument is given it should return `true` if at least one of the elements in the array is `true`. Otherwise, it returns `false`. Line 63 does not address this as you think. [See comment]
- when a pattern other than Regex or a Class is given it should `false` if none of the collection matches the pattern and `true` otherwise

You should consider returning false by default and return true if an element meets criteria. This would solve a lot of your problems

File: enum.rb
Comment (mcrd25) [reply enum.rb:70]: You'd need to also pass a second &block parameter along with pattern in `my_none` and pass it to your negate `my_any` to handle blocks

`def my_none?(pattern = nil, &block)
    !my_any?(pattern, &block)
  end`

**Review by mcrd25 — CHANGES_REQUESTED** [review]
# Changes Required 🤓 

Hi @noomdalv 👋,

Great job with implementation so far! 👏 
You've managed to pass all but 7 tests of your Enumerables implementation. I've noticed you've had this review open for quite some time. Let's make this the last iteration and work together to get your methods working as they should.

Methods that are 💯 
- `my_each`
- `my_each_with_index`
- `my_select`
- `my_all?`
- `my_map`
- `my_count`
Methods that need work:
- `my_any?`
- `my_none?`
- `my_inject`

Address my review comments for the methods that need refactoring. You can get more info about these methods with the following link: https://ruby-doc.org/core-2.6.3/Enumerable.html
This would show you how the original methods handle different arguments etc and you can mimic the behavior in your methods.

I know this has been mentioned previously, but please remember to use [Git Flow](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59882084/) in next project.
***How Git Flow Works***
  - You'd have the `master` branch purely for production purposes 
  - the `development` branch would act as the main working branch where you create merge requests from `feature` branches to `development` for reviews on these requests
  - If you are requesting a review for a project without multiple milestones simply have one main `feature` branch (appropriately named) that you can then branch off from and merge to accordingly as you complete your project. 
  - After you have completed your project and it's ready for your first review you can then request a code review form your main `feature` branch to `development` branch
  - Once the pull request is approved you can then merge this pull request
  - You would then merge the `development` branch to `master` branch 

*I really like your implementations so far, Please address changes and let's bucket this one! 💪*


***Please makes changes and submit another review request.
If you have any questions, let me know here or on slack and I will be glad to clarify with you. You can also reach me and connect with me on my social media or email.***

*Happy coding 😄,*
**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

---
*P.S. all requested changes not marked by [OPTIONAL] or are not positive comments are **REQUIRED***

File: enum.rb
Comment (mcrd25) [reply enum.rb:70]: NOTE: This is not necessary or original `my_any` but for `my_none`

**Review by Oghenebrume50 — CHANGES_REQUESTED** [review]
Hello coder, you have made real progress, :clap: I cannot commend yo enough sincerely, there are just two failing test that should be looked out, please see them below. 
[OPTIONAL] You might want to simplify this very method(my_inject) but first get it to work.

**Failing test** 
my_inject combines all elements of enum by applying a binary operation, specified by a block
my_inject, when a symbol is specified, combines each element of the collection by applying the symbol as a named method.

Please make those changes and request for a re-review :pray:  Cheers :clinking_glasses: to growth.

**Review by Ispirett — CHANGES_REQUESTED** [review]
Hey, @noomdalv  you have a lot of patience that's a very good character trait to have. image. 

image

I would really love to pass the project :+1:  but that's just a 2 bug :bug:  to fix

I tried this against your my inject method and they produce different results.

 ` array.my_inject { |p,e| p + e }  ==  array.inject { |p,e| p + e }`
 `array.my_inject(:+) ==  array.inject(:+)`

**Your almost the dont give up I'm also sure you have learn a lot.**

**Comment by noomdalv** [comment]
fixed

array.my_inject { |p,e| p + e } == array.inject { |p,e| p + e }
true

array.my_inject(:+) == array.inject(:+)
true

File: enum.rb
Comment (Ispirett) [inline enum.rb:112]: The array here was never declared

File: enum.rb
Comment (Ispirett) [inline enum.rb:120]: The array here was never declared

**Review by Ispirett — CHANGES_REQUESTED** [review]
Great job @noomdalv  with changes  :100: 
image. 
But there is not this error that results from the my_inject method 

 **undefined local variable or method `array'**


## Git/flow reminder.

Friendly reminder: You should use Git Flow in this project. However, changing it now would take too much time.
Please make sure you use Git Flow in the next project. You will find instructions in the Patwright lesson description.

**Comment by noomdalv** [comment]
sry, forgot to add the array declaration again while trying to fix the
other errors, its working now with all the test cases(they are commented)

**Review by Ispirett — APPROVED** [review]
Really Great work @noomdalv  you can merge now!!!  
image

image
## Git/flow reminder.

Friendly reminder: You should use Git Flow in this project. However, changing it now would take too much time.
Please make sure you use Git Flow in the next project. You will find instructions in the Patwright lesson description.

**Pull request merged**
